<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Mathematics of single-cell omics - Diffusion distances</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Diffusion distances</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="load-example-data" class="level4">
<h4 class="anchored" data-anchor-id="load-example-data">Load example data</h4>
<p>Before getting into the topic, we load the usual example data and performing standard preprocessing</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( tidyverse )</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( Matrix )</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( sparseMatrixStats )</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( Seurat ) })</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ReadMtx</span>( <span class="st">"~/Downloads/ifnagrko/ifnagrko_raw_counts.mtx.gz"</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/ifnagrko/ifnagrko_obs.csv"</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/ifnagrko/ifnagrko_var.csv"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">cell.sep=</span><span class="st">","</span>, <span class="at">feature.sep=</span><span class="st">","</span>, <span class="at">skip.cell=</span><span class="dv">1</span>, <span class="at">skip.feature=</span><span class="dv">1</span>, </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">mtx.transpose=</span><span class="cn">TRUE</span>) <span class="ot">-&gt;</span> count_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>count_matrix <span class="sc">%&gt;%</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">CreateSeuratObject</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">RunPCA</span>( <span class="at">npcs=</span><span class="dv">20</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">FindNeighbors</span>( <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">FindClusters</span>( <span class="at">resolution=</span><span class="fl">0.5</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">RunUMAP</span>( <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span> ) <span class="ot">-&gt;</span> seu</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  Apoe, Aldoc, Sparcl1, Sdc4, Ptn, Cmtm5, Glul, Gpr37l1, Fxyd1, Atp1b2 
       S100a1, Slc4a4, Slc1a3, Prxl2a, F3, Itm2b, Mt1, Rgcc, Prdx6, Sfxn5 
       Sat1, Scrg1, Dbi, Hes5, Luzp2, Plaat3, Pla2g7, Sash1, Plpp3, Sparc 
Negative:  Tubb5, Sox11, Tubb3, Stmn1, Jpt1, Hmgb3, Ptma, Sox4, Dlx2, Cd24a 
       Igfbpl1, Dlx6os1, Map1b, Stmn2, Abracl, Tmsb4x, Lmnb1, Cdca7, Ccnd2, Elavl4 
       Cdk4, Dcx, Arx, Uchl1, EYFP, Celf4, Dlx5, Nrxn3, H1fx, Hmgn2 
PC_ 2 
Positive:  Ctss, C1qc, Laptm5, Csf1r, Trem2, C1qa, Cx3cr1, C1qb, Tyrobp, Ly86 
       Fcer1g, Siglech, Selplg, Fcrls, Tmem119, Fcgr3, Apbb1ip, Unc93b1, Cd53, Lpcat2 
       Spi1, Pld4, Olfml3, Irf8, Ctsh, Aif1, Cd300c2, Fyb, Otulinl, Mylip 
Negative:  Rorb, Cldn10, Clu, Mt3, Ntsr2, Mfge8, S1pr1, Id4, Slc1a2, Acsl6 
       Plpp3, Sox9, Ddah1, Bcan, Cxcl14, Btbd17, Mlc1, Cspg5, Fjx1, Aqp4 
       Ntm, Acsl3, Gabrb1, Tspan7, Lsamp, Chst2, Mt2, Lhx2, Slc39a12, Glud1 
PC_ 3 
Positive:  Atp1a3, Camk2b, Snhg11, Syt1, Nrip3, Kcnj4, Scg2, Snap25, Dnm1, Pcp4 
       Icam5, Ndrg4, Eef1a2, Eno2, Ano3, Ryr2, Arpp21, Ptk2b, Gng4, Kcna4 
       Penk, Slc4a10, Snca, Gad1, Rprml, Grin2a, C1qtnf4, Shisa8, Camk2a, Kcnb2 
Negative:  Hmgb2, Top2a, Pbk, Birc5, Mki67, Cdk1, Cdca8, Spc24, Cenpf, Spc25 
       Prc1, Rrm2, Mdk, Nusap1, Tpx2, Cdca3, Knl1, Ckap2l, Esco2, Aurkb 
       Cenpm, Ccna2, Bub1, Cks2, Kif11, Hist1h3c, Hist1h1b, Hmmr, Pclaf, Fbxo5 
PC_ 4 
Positive:  C1qc, C1qa, Ctss, Trem2, Csf1r, C1qb, Cx3cr1, Laptm5, Fcer1g, Tyrobp 
       Ly86, Siglech, Selplg, Fcrls, Fcgr3, Hexb, Spi1, Cd53, Itgb5, Pld4 
       Ptgs1, Cd300c2, Aif1, Irf8, Fyb, Itgam, Cyth4, Ltc4s, Otulinl, Cd37 
Negative:  Frzb, Apod, Npy, Plp1, Vtn, Foxd3, Wnt6, Nr2f2, Edil3, Sox10 
       Gsn, Matn4, Fbln2, Aspa, Aqp1, Igf1, Plat, Lpar1, Igfbp4, Erbb3 
       Fabp7, Plppr4, Ptgds, Col23a1, Alx3, Hey2, Cd59a, Fam3c, Scd1, Mybpc1 
PC_ 5 
Positive:  Stmn2, Igfbpl1, Cd24a, Nrep, Sox4, Map1b, Stmn4, Tubb3, Shtn1, Dlx6os1 
       Dcx, Ly6h, Sox11, Jpt1, Mpped2, Stmn1, Plxna4, Pbx3, Elavl4, Uchl1 
       Runx1t1, Cald1, Foxp2, Dlx2, Gad2, Celf4, Pfn2, Dlx5, Sp8, Tubb5 
Negative:  Top2a, Pbk, Birc5, Mki67, Spc25, Cdk1, Prc1, Nusap1, Spc24, Esco2 
       Tpx2, Knl1, Aurkb, Cenpf, Cdca8, Ckap2l, Kif11, Cdca3, Hist1h3c, Hmmr 
       Ccna2, Bub1, Incenp, Hist1h2af, Ndc80, Cit, Fbxo5, Kif4, Sgo1, Kif22 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 18302
Number of edges: 616069

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9191
Number of communities: 19
Elapsed time: 4 seconds</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:32:17 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:32:17 Read 18302 rows and found 20 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:32:17 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>15:32:17 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
15:32:20 Writing NN index file to temp file /tmp/Rtmpph7jNv/file504c155b54113
15:32:20 Searching Annoy index using 1 thread, search_k = 3000
15:32:28 Annoy recall = 100%
15:32:28 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
15:32:29 Initializing from normalized Laplacian + noise (using RSpectra)
15:32:33 Commencing optimization for 200 epochs, with 776086 positive edges
15:32:43 Optimization finished</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>( seu, <span class="at">label=</span><span class="cn">TRUE</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="nearest-neighbors" class="level4">
<h4 class="anchored" data-anchor-id="nearest-neighbors">Nearest neighbors</h4>
<p>In this lesson, we will mainly work with the nearest neighbor data. Seurat has already calculated this but we do this again here:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>FNN<span class="sc">::</span><span class="fu">get.knn</span>( <span class="fu">Embeddings</span>( seu, <span class="st">"pca"</span> ), <span class="at">k=</span><span class="dv">15</span> ) <span class="ot">-&gt;</span> nn</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( nn<span class="sc">$</span>nn.index)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1]  [,2]  [,3]  [,4]  [,5]  [,6]  [,7]  [,8]  [,9] [,10] [,11] [,12]
[1,]  2504  4613  2817  2697  4423  3301 11993 17335   782  7322  6832   710
[2,] 12412 11120  9817  6433 11238 13771  1477 11146 11275 16123 16751  5537
[3,] 10073  9512  3943  7030  8300 10489  5189  7207  3050 10479  5624  9527
[4,] 12407  2608 16034  8035  1662  4011  3937 10428 16823  5610 16173  3883
[5,] 10060 10135  3874  5333  4318  4437  7573 16224 11594  6338  8306 17955
[6,] 13639 11873 16237 12423 17155  2140  8348  1213  5423  2691 12754  5443
     [,13] [,14] [,15]
[1,]  4421  2209  5197
[2,]  9427  1947  8633
[3,]  3086  1516  8733
[4,]  1858  4214 16949
[5,] 14159  5662 13930
[6,]   185  3376   273</code></pre>
</div>
</div>
</section>
<section id="adjacency-matrix" class="level4">
<h4 class="anchored" data-anchor-id="adjacency-matrix">Adjacency matrix</h4>
<p>We next construct the adjacency matrix of the undirected nearest neighbor graph. To connect each vertex to its <span class="math inline">\(i\)</span>th nearest neighbor, we need the following edges:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>ncells <span class="ot">&lt;-</span> <span class="fu">ncol</span>(seu)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>( <span class="at">vertex_A=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">vertex_B=</span>nn<span class="sc">$</span>nn.index[,i] ) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     vertex_A vertex_B
[1,]        1     2817
[2,]        2     9817
[3,]        3     3943
[4,]        4    16034
[5,]        5     3874
[6,]        6    16237</code></pre>
</div>
</div>
<p>We can use this as indices of the matrix cells we want to set to one and thus construct a sparse matrix</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sparseMatrix</span>(  <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span>nn<span class="sc">$</span>nn.index[,i], <span class="at">x=</span><span class="dv">1</span>, <span class="at">dims=</span><span class="fu">c</span>(ncells,ncells) ) <span class="sc">%&gt;%</span> <span class="fu">summary</span>() <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>18302 x 18302 sparse Matrix of class "dgCMatrix", with 18302 entries
      i j x
1  2697 1 1
2  2817 1 1
3  7322 1 1
4 11993 1 1
5  9427 2 1
6  9817 2 1</code></pre>
</div>
</div>
<p>Adding up on such matrix for i running from 1 to 15 gives us the adjacency matrix</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>adjm <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>( <span class="at">i=</span><span class="fu">integer</span>(), <span class="at">j=</span><span class="fu">integer</span>(), <span class="at">x=</span><span class="fu">numeric</span>(), <span class="at">dims=</span><span class="fu">c</span>(ncells,ncells) ) <span class="co"># zero matrix</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>( i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(nn<span class="sc">$</span>nn.index) ) {</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>   adjm <span class="ot">&lt;-</span> adjm <span class="sc">+</span> <span class="fu">sparseMatrix</span>(  <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span>nn<span class="sc">$</span>nn.index[,i], <span class="at">x=</span><span class="dv">1</span>, <span class="at">dims=</span><span class="fu">c</span>(ncells,ncells) ) }</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(adjm) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>18302 x 18302 sparse Matrix of class "dgCMatrix", with 274530 entries
     i j x
1  174 1 1
2  710 1 1
3  782 1 1
4 1121 1 1
5 1769 1 1
6 2504 1 1</code></pre>
</div>
</div>
<p>We make the matrix symmetric</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>adjm <span class="ot">&lt;-</span>  adjm <span class="sc">+</span> <span class="fu">t</span>(adjm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, some matrix entries have become 2 rather than 1. We set everything back to 1:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>adjm<span class="sc">@</span>x[] <span class="ot">&lt;-</span> <span class="dv">1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we have an adjacency matrix for our nearest neighbor graph.</p>
</section>
<section id="random-walk" class="level4">
<h4 class="anchored" data-anchor-id="random-walk">Random walk</h4>
<p>We now define a random walk on our graph as follows: A “walker” (or: “token”) starts at a vertex <span class="math inline">\(i\)</span>. In each time step, it choses one of the vertex’s neighbors at random and moves there. What is the probability of the walker being on vertex <span class="math inline">\(j\)</span> after <span class="math inline">\(\ell\)</span> steps?</p>
<p>We represent the walker being at vertex <span class="math inline">\(i\)</span> with the unit vector in direction <span class="math inline">\(i\)</span>, i.e., the vector <span class="math inline">\(\vec e_i\)</span>, with a 1 at component <span class="math inline">\(i\)</span> and zero elsewhere. The transition matrix <span class="math inline">\(T\)</span> with elements <span class="math inline">\(T_{ij}\)</span> tells us the probability of the walker moving to vertex <span class="math inline">\(j\)</span> in a step if it was before at vertex <span class="math inline">\(i\)</span>:</p>
<p><span class="math display">\[ T_{ij} = A_{ij} \Big/ \sum_{j'}{A_{ij'}}. \]</span> The division normalizes the probabiliities by dividing by the number of neighbors that the walker can chose from.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>trm <span class="ot">&lt;-</span> adjm <span class="sc">/</span> <span class="fu">rowSums</span>(adjm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Check normalization:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowSums</span>(trm) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1 1 1 1 1 1</code></pre>
</div>
</div>
<p>To try this out, we pick a cell close to the point (-6,6) in the UMAP:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>cell <span class="ot">&lt;-</span> <span class="fu">which.min</span>( ( <span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>)[,<span class="dv">1</span>] <span class="sc">+</span> <span class="dv">6</span> )<span class="sc">^</span><span class="dv">2</span> <span class="sc">+</span> ( <span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>)[,<span class="dv">2</span>] <span class="sc">-</span> <span class="dv">6</span> )<span class="sc">^</span><span class="dv">2</span> )</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>cell</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>CGGGCATGTTGGGAAC-WT_4mo_2021_001 
                           14741 </code></pre>
</div>
</div>
<p>Here’s a UMAP plot of this cell and it’s neighbors:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">w =</span> <span class="fu">case_when</span>(</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">row_number</span>() <span class="sc">==</span> cell <span class="sc">~</span> <span class="st">"cell"</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  adjm[cell,] <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"neighbor"</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"other"</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>)) <span class="sc">%&gt;%</span> </span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>w ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>( <span class="at">values=</span><span class="fu">c</span>(<span class="st">"darkgreen"</span>,<span class="st">"magenta"</span>,<span class="st">"#00000006"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s perform 10 steps. We start with a sparse vector with a single 1 at the chosen cell’s index and 0 elsewhere, then multiply this 10 times with <span class="math inline">\(T\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>u <span class="ot">&lt;-</span> <span class="fu">sparseVector</span>( <span class="at">i=</span>cell, <span class="at">x=</span><span class="dv">1</span>, <span class="at">length=</span>ncells )</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>( i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span> )</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  u <span class="ot">&lt;-</span> u <span class="sc">%*%</span> trm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We first check whether <code>u</code> is still normalized:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(u)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<p>Here’s a plot of <span class="math inline">\(\vec{u}=\vec{e}_i^\top T^{10}\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">u =</span> <span class="fu">as.vector</span>(u) ) <span class="sc">%&gt;%</span> </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>u ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">direction=</span><span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="exponantiating-the-transition-matrix" class="level4">
<h4 class="anchored" data-anchor-id="exponantiating-the-transition-matrix">Exponantiating the transition matrix</h4>
<p>Calculating <span class="math inline">\(\ell\)</span> steps by repeated multiplication is wasteful. We should use a matrix exponential.</p>
<p>As preparation for this, we define the diagonal “degree matrix” <span class="math inline">\(D\)</span>, that contains the vertex degrees: <span class="math display">\[ D_{ij} = \delta_{ij} \sum_{j'} A_{ij'}. \]</span></p>
<p>Now, we have: <span class="math display">\[ T = D^{-1}A.\]</span></p>
<p><span class="math inline">\(T\)</span> is a row-stochastic matrix, i.e., its values are all non-negative and its rows sum to 1.</p>
<p>The probability mass vector for a walker starting at vertex <span class="math inline">\(i\)</span> after one step is <span class="math inline">\(\vec{e}_i^\top T\)</span>, and after <span class="math inline">\(\ell\)</span> steps, <span class="math inline">\(\vec{e}_i^\top T^\ell\)</span>.</p>
<p>To calculate <span class="math inline">\(T^\ell\)</span>, we will need the eigendecomposition of the symmetrized transition matrix <span class="math inline">\(\tilde T = D^{-1/2} T D^{-1/2}\)</span>: <span class="math display">\[ \tilde T = U\Lambda U^\top,\]</span> with the columns of <span class="math inline">\(U\)</span> containing the eigenvectors of <span class="math inline">\(\tilde T\)</span> and the diagonal matrix <span class="math inline">\(\Lambda\)</span> containing the eigenvalues.</p>
<p>With this, we get <span class="math display">\[ \begin{align}
T^\ell &amp;= \left(D^{-1} A\right)^\ell \\
&amp;= D^{-1/2} \left(D^{-1/2} A D^{-1/2}\right)^\ell D^{1/2} \\
&amp;= D^{-1/2} \left(U \Lambda U^T\right)^\ell D^{1/2} \\
&amp;= \underbrace{D^{-1/2} U \Lambda^\ell}_{=X_\ell} U^T D^{1/2}.
\end{align} \]</span></p>
<p>We construct <span class="math inline">\(D\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>degdiag <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>( <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">x=</span><span class="fu">rowSums</span>(adjm) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We also write down <span class="math inline">\(D^{-1}\)</span>, <span class="math inline">\(D^{1/2}\)</span> and <span class="math inline">\(D^{-1/2}\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>invdegdiag     <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>( <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">x=</span><span class="dv">1</span><span class="sc">/</span><span class="fu">rowSums</span>(adjm) )</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>sqrtdegdiag    <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>( <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">x=</span><span class="fu">sqrt</span>(<span class="fu">rowSums</span>(adjm)) )</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>invsqrtdegdiag <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>( <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">x=</span><span class="dv">1</span><span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">rowSums</span>(adjm)) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we get the eigensystem of <span class="math inline">\(\tilde T\)</span>, requesting the 100 eigenvalues that are largest by magnitude:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>eigtrm <span class="ot">&lt;-</span> RSpectra<span class="sc">::</span><span class="fu">eigs_sym</span>( invsqrtdegdiag <span class="sc">%*%</span> adjm <span class="sc">%*%</span> invsqrtdegdiag, <span class="at">k=</span><span class="dv">100</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can calculate <span class="math inline">\(\vec e_i^\top T^\ell\)</span> as follows:</p>
<p>We calculate first <span class="math inline">\(X_\ell = D^{-1/2} U \Lambda^\ell\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>invsqrtdegdiag <span class="sc">%*%</span> eigtrm<span class="sc">$</span>vectors <span class="sc">%*%</span> <span class="fu">diag</span>( eigtrm<span class="sc">$</span>values<span class="sc">^</span><span class="dv">10</span> ) <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>() <span class="ot">-&gt;</span> x10</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(x10)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18302   100</code></pre>
</div>
</div>
<p>Note that the matrix <span class="math inline">\(X_\ell\)</span> has been trimmed to only the first 50 columns. That is ok because the factor <span class="math inline">\(\Lambda^\ell\)</span> gets small quickly and therefore, the columns stay close to zero once we get past the first few on the left:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( <span class="fu">colSums</span>(x10<span class="sc">^</span><span class="dv">2</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Of course, this only works for <span class="math inline">\(\ell \gg 1\)</span> and with <span class="math inline">\(\ell=10\)</span> we may have only just enough steps for the approximation becoming valid.</p>
<p>We take the row of <span class="math inline">\(X_\ell\)</span> that corresponds to our cell and multiply this with <span class="math inline">\(U^\top D^{1/2}\)</span> to get <span class="math inline">\(e_i^\top X_\ell U^\top D^{1/2}=e_i^\top T^\ell\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>x10[cell,] <span class="sc">%*%</span> <span class="fu">t</span>(eigtrm<span class="sc">$</span>vectors) <span class="sc">%*%</span> sqrtdegdiag <span class="sc">%&gt;%</span> <span class="fu">as.vector</span>() <span class="ot">-&gt;</span> u10</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is a plot of that vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">u =</span> u10 ) <span class="sc">%&gt;%</span> </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>u ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">direction=</span><span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This looks quite similar as the plot we made before.</p>
<p>Here’s a comparison of the two results</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( u, u10, <span class="at">asp=</span><span class="dv">1</span>, <span class="at">cex=</span>.<span class="dv">2</span>, <span class="at">xlab=</span><span class="st">"exact calculation"</span>, <span class="at">ylab=</span><span class="st">"using top 100 eigenvectors"</span> )</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>( <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">col=</span><span class="st">"#00000020"</span> )</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>( <span class="at">h=</span><span class="dv">0</span>, <span class="at">v=</span><span class="dv">0</span>, <span class="at">col=</span><span class="st">"#00000020"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="diffusion-distances" class="level4">
<h4 class="anchored" data-anchor-id="diffusion-distances">Diffusion distances</h4>
<p>We now define a new distance metrix for our cells. Intuitively: To quantify the distance between two cells <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, we start random walks at both cells, evolve them for <span class="math inline">\(\ell\)</span> steps, and then ask about the overlap between the resulting probability vectors.</p>
<p>We might therefore use <span class="math display">\[ \left\| e_i^\top T^\ell - e_j
^\top T^\ell \right\|_2, \]</span> i.e., the Euclidean distance between rows <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> of <span class="math inline">\(T^\ell\)</span> as the distance between cells <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>.</p>
<p>Note that the number of steps, <span class="math inline">\(\ell\)</span>, selects a “length scale” at which the that distance is informative.</p>
<p>There is a practical difficulty in using this definition, though: When multiplying the small <span class="math inline">\(n\times k\)</span> matrix <span class="math inline">\(X_\ell\)</span> with the transpose of the <span class="math inline">\(n\times k\)</span> eigenvector matrix <span class="math inline">\(U\)</span> (with <span class="math inline">\(n\)</span> being the number of cells and <span class="math inline">\(k\)</span> being the number of eigenvectors that have been calculated) our data blows up to an unwieldy <span class="math inline">\(n\times n\)</span> matrix.</p>
<p>We can avoid this by using the rows of <span class="math inline">\(X_\ell\)</span> instead of the rows of <span class="math inline">\(T^\ell\)</span>, and therefore define:</p>
<p>The <em><span class="math inline">\(\ell\)</span>-steps diffusion distance</em> between cells <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> is <span class="math display">\[ d_{\ell,ij} = \left\| e_i^\top X_\ell - e_j^\top X_\ell \right\|_2, \]</span></p>
<p>Note that this is also <span class="math display">\[ d_{\ell,ij} = \left\| \left( e_i^\top T^\ell  - e_i^\top T^\ell\right) D^{-1/2} \right\|_2, \]</span> i.e., the components of the resulting probability vectors get reweighted by <span class="math inline">\(D^{-1/2}\)</span> before calculating the norm of the difference. If we accept this (somewhat unmotivated) reweighting, because it does not change much (as the vertex degrees do not differ that much from each other), we have a computationally efficient way of calculating <span class="math inline">\(d_{\ell,ij}\)</span>: All we need is <span class="math inline">\(X_\ell\)</span>.</p>
<p>By calculating the Euclidean distance of every row of <span class="math inline">\(X_\ell\)</span> to the row for our selected cell, we get the <span class="math inline">\(\ell\)</span>-step diffusion distance of that cell to all other cells:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">d =</span> <span class="fu">sqrt</span>( <span class="fu">rowSums</span>( <span class="fu">t</span>( <span class="fu">t</span>(x10) <span class="sc">-</span> x10[cell,] )<span class="sc">^</span><span class="dv">2</span> ) ) ) <span class="sc">%&gt;%</span> </span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(<span class="sc">-</span>d) <span class="sc">%&gt;%</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>d ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">direction=</span><span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Below, this plot is repeated for several values of <span class="math inline">\(\ell\)</span>. (But remember that our approximation error is larger for <span class="math inline">\(\ell \lesssim 10\)</span>.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>( l <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">10</span>, <span class="dv">30</span>, <span class="dv">100</span>, <span class="dv">300</span>, <span class="dv">1000</span> ) ) {</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  invsqrtdegdiag <span class="sc">%*%</span> eigtrm<span class="sc">$</span>vectors <span class="sc">%*%</span> <span class="fu">diag</span>( eigtrm<span class="sc">$</span>values<span class="sc">^</span>l ) <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>() <span class="ot">-&gt;</span> xm</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>( <span class="at">d =</span> <span class="fu">sqrt</span>( <span class="fu">rowSums</span>( <span class="fu">t</span>( <span class="fu">t</span>(xm) <span class="sc">-</span> xm[cell,] )<span class="sc">^</span><span class="dv">2</span> ) ) ) <span class="sc">%&gt;%</span> </span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(<span class="sc">-</span>d) <span class="sc">%&gt;%</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>    ggplot <span class="sc">+</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>d ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_viridis_c</span>(<span class="at">direction=</span><span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>( <span class="fu">sprintf</span>( <span class="st">"%d steps"</span>, l ) ) )</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-26-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-26-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-26-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-26-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-26-6.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="pseudotime" class="level4">
<h4 class="anchored" data-anchor-id="pseudotime">Pseudotime</h4>
<p>We now want to define a pseudotime along the lineage, from cluster 0 to cluster 2 and on to 7.</p>
<p>We define a start cell and an end cell, by taking our cell (which is arguably somewhere in the middle) and find the cell with the largest distance to it within cluster 0 and 7, respectivelty.</p>
<p>We will work with <span class="math inline">\(\ell=300\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>l <span class="ot">&lt;-</span> <span class="dv">300</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>xm <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>( invsqrtdegdiag <span class="sc">%*%</span> eigtrm<span class="sc">$</span>vectors <span class="sc">%*%</span> <span class="fu">diag</span>( eigtrm<span class="sc">$</span>values<span class="sc">^</span>l ) )</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="co"># distances to "intermediate" cell:</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">sqrt</span>( <span class="fu">rowSums</span>( <span class="fu">t</span>( <span class="fu">t</span>(xm) <span class="sc">-</span> xm[cell,] )<span class="sc">^</span><span class="dv">2</span> ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is the distance of the cells in cluster 0 (the astrocyte / neuronal stem cell cluster, where the lineage starts) to our intermediate cell:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>( d[seu<span class="sc">$</span>seurat_clusters<span class="sc">==</span><span class="dv">0</span>], <span class="dv">100</span> )</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="fl">0.0013</span>, <span class="at">col=</span><span class="st">"orange"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>It seems reasonable to assume that the cells at the steep cliff are the actual start of the lineage and the cells further out or some outliers (perhaps cells that started into another direction)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>start_cell <span class="ot">&lt;-</span> <span class="fu">which.min</span>( ( d <span class="sc">*</span> (seu<span class="sc">$</span>seurat_clusters<span class="sc">==</span><span class="dv">0</span>) <span class="sc">-</span> <span class="fl">0.0013</span> )<span class="sc">^</span><span class="dv">2</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s also find a reasonable “end cell” in the neuron cluster 7:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>( d[seu<span class="sc">$</span>seurat_clusters<span class="sc">==</span><span class="dv">7</span>], <span class="dv">100</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here’ let’s simply take the last cell:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>end_cell <span class="ot">&lt;-</span> <span class="fu">which.max</span>( d <span class="sc">*</span> (seu<span class="sc">$</span>seurat_clusters<span class="sc">==</span><span class="dv">7</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, get distances to these two cells:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>dist_to_start <span class="ot">&lt;-</span> <span class="fu">sqrt</span>( <span class="fu">rowSums</span>( <span class="fu">t</span>( <span class="fu">t</span>(xm) <span class="sc">-</span> xm[start_cell,] )<span class="sc">^</span><span class="dv">2</span> ) )</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>dist_to_end <span class="ot">&lt;-</span> <span class="fu">sqrt</span>( <span class="fu">rowSums</span>( <span class="fu">t</span>( <span class="fu">t</span>(xm) <span class="sc">-</span> xm[end_cell,] )<span class="sc">^</span><span class="dv">2</span> ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>( </span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  dist_to_start, dist_to_end, </span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster=</span>seu<span class="sc">$</span>seurat_clusters ) <span class="sc">%&gt;%</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="fu">case_when</span>(</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>    cluster <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span> ) <span class="sc">~</span> <span class="st">"lineage_straight"</span>,</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>    cluster <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">6</span>, <span class="dv">11</span> )  <span class="sc">~</span> <span class="st">"lineage_cycle"</span>,</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"other"</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>dist_to_end, <span class="at">y=</span>dist_to_start, <span class="at">col=</span>type ), <span class="at">size=</span>.<span class="dv">1</span> ) <span class="sc">+</span></span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s rotate this plot by 45°:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>( </span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  dist_to_start, dist_to_end, </span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster=</span>seu<span class="sc">$</span>seurat_clusters ) <span class="sc">%&gt;%</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="fu">case_when</span>(</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>    cluster <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span> ) <span class="sc">~</span> <span class="st">"lineage_straight"</span>,</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>    cluster <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">6</span>, <span class="dv">11</span> )  <span class="sc">~</span> <span class="st">"lineage_cycle"</span>,</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"other"</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( </span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> dist_to_start <span class="sc">-</span> dist_to_end, </span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> dist_to_start <span class="sc">+</span> dist_to_end, <span class="at">col=</span>type ), <span class="at">size=</span>.<span class="dv">1</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Our new x-axis should be suitable as pseudotime:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>pseudotime <span class="ot">&lt;-</span> dist_to_start <span class="sc">-</span> dist_to_end</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s plot this into the UMAP:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="fu">add_column</span>( pseudotime ) <span class="sc">%&gt;%</span> </span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>pseudotime ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>( <span class="at">colours=</span>rje<span class="sc">::</span><span class="fu">cubeHelix</span>(<span class="dv">100</span>,<span class="at">r=</span><span class="dv">4</span>) ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Before we dive into this, let’s also plot the <code>y axis of</code>dist+to+start+dfist+to_end`:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">add_column</span>( <span class="at">distsum =</span> dist_to_start <span class="sc">+</span> dist_to_end ) <span class="sc">%&gt;%</span> </span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>distsum ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>( <span class="at">midpoint =</span> .<span class="dv">01</span>, <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,.<span class="dv">02</span>), <span class="at">oob=</span>scales<span class="sc">::</span>oob_squish )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Comparing with the plot above, our pseudotime is valid for the red, white and perhaps the very faintly blue regions.</p>
</section>
<section id="comaprison-to-principal-curve" class="level3">
<h3 class="anchored" data-anchor-id="comaprison-to-principal-curve">Comaprison to principal curve</h3>
<p>Let’s recreate the principal curve that we used before for pseudotime.</p>
<p>As before, we calculate the principal curve using the cells in the lineage as input, without the cells from the two cycling cluster, and use smoothing splines with 10 degrees of freedom.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>princurve<span class="sc">::</span><span class="fu">principal_curve</span>(</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>)[ seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span> ), ],</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">df =</span> <span class="dv">10</span>, <span class="at">approx_points=</span><span class="dv">1000</span> ) <span class="ot">-&gt;</span> prc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We again project the remaining cells onto outo the cells on the curve and assign to them the same pseudotime as the one of the nearest point on the curve:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>FNN<span class="sc">::</span><span class="fu">get.knnx</span>( prc<span class="sc">$</span>s, <span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>), <span class="at">k=</span><span class="dv">1</span> ) <span class="ot">-&gt;</span> prc_nn</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>pseudotime_prc <span class="ot">&lt;-</span> prc<span class="sc">$</span>lambda[ prc_nn<span class="sc">$</span>nn.index ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is the principal-curve-based pseudotime:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="fu">add_column</span>( pseudotime_prc ) <span class="sc">%&gt;%</span> </span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>pseudotime_prc ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>() <span class="sc">+</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>( <span class="at">colours=</span>rje<span class="sc">::</span><span class="fu">cubeHelix</span>(<span class="dv">100</span>,<span class="at">r=</span><span class="dv">4</span>) ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can also compare the two:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt_diffusion =</span> pseudotime,</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt_prc =</span> <span class="sc">-</span>pseudotime_prc,</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster =</span> seu<span class="sc">$</span>seurat_clusters</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>(</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="fu">case_when</span>(</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>    cluster <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span> ) <span class="sc">~</span> <span class="st">"lineage_straight"</span>,</span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>    cluster <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">6</span>, <span class="dv">11</span> )  <span class="sc">~</span> <span class="st">"lineage_cycle"</span>,</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"other"</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>pt_diffusion, <span class="at">y=</span>pt_prc, <span class="at">col=</span>type ), <span class="at">size=</span>.<span class="dv">2</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="diffusion-space" class="level3">
<h3 class="anchored" data-anchor-id="diffusion-space">Diffusion space</h3>
<p>The space of <span class="math inline">\(X_\ell\)</span> can also be interpreted as an alternative to the feature space, called “diffusion space”. Using the first few components provides a dimension redution, the “diffusion map”.</p>
<p>For this to work well, our neighborhood graph should be connected.</p>
<p>Ours has two connection components, however, as is evident from the fact that our transition matrix has two unit eigenvalues:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( eigtrm<span class="sc">$</span>values )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1.0000000 1.0000000 0.9998821 0.9993076 0.9991653 0.9984226</code></pre>
</div>
</div>
<p>To make our live easier, let’s reduce the data to only the lineage cells</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>in_lineage <span class="ot">&lt;-</span> seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">0</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span>, <span class="dv">6</span>, <span class="dv">11</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Subset the adjacency matrix to these and recalculate the transition matrix and its spectrum</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>adjml <span class="ot">&lt;-</span> adjm[ in_lineage, ][ , in_lineage ]</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>ncells <span class="ot">&lt;-</span> <span class="fu">nrow</span>(adjml)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>invsqrtdegdiag <span class="ot">&lt;-</span> <span class="fu">sparseMatrix</span>( <span class="at">i=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">j=</span><span class="dv">1</span><span class="sc">:</span>ncells, <span class="at">x=</span><span class="dv">1</span><span class="sc">/</span><span class="fu">sqrt</span>(<span class="fu">rowSums</span>(adjml)) )</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>eigtrm <span class="ot">&lt;-</span> RSpectra<span class="sc">::</span><span class="fu">eigs_sym</span>( invsqrtdegdiag <span class="sc">%*%</span> adjml <span class="sc">%*%</span> invsqrtdegdiag, <span class="at">k=</span><span class="dv">10</span> )</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>x300 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>( invsqrtdegdiag <span class="sc">%*%</span> eigtrm<span class="sc">$</span>vectors <span class="sc">%*%</span> <span class="fu">diag</span>( eigtrm<span class="sc">$</span>values<span class="sc">^</span><span class="dv">300</span> ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>( x300[,<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span> ] ) <span class="sc">%&gt;%</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">add_column</span>( <span class="at">cluster =</span> seu<span class="sc">$</span>seurat_clusters[in_lineage] ) <span class="sc">%&gt;%</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>V1, <span class="at">y=</span>V2, <span class="at">col=</span>cluster ), <span class="at">size=</span>.<span class="dv">1</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `x` argument of `as_tibble.matrix()` must have unique column names if
`.name_repair` is omitted as of tibble 2.0.0.
ℹ Using compatibility `.name_repair`.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>For comparison, the UMAP with the same c</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>( <span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>)[in_lineage,] ) <span class="sc">%&gt;%</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="fu">add_column</span>( <span class="at">cluster =</span> seu<span class="sc">$</span>seurat_clusters[in_lineage] ) <span class="sc">%&gt;%</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>cluster ), <span class="at">size=</span>.<span class="dv">1</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="graph_diffusion_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<p>The idea of diffusion distances and diffusion maps has been introduced in this paper:</p>
<ul>
<li>Coifman and Lafon (2006): <em>Diffusion Maps</em>. Applied and Computational Harmonic Analysis, Vol. 21, Pages 5-30. <a href="https://doi.org/10.1016/j.acha.2006.04.006">doi:10.1016/j.acha.2006.04.006</a></li>
</ul>
<p>Applying these ideas to single-cell data is explored in</p>
<ul>
<li><p>Haghverdi, Büttner, Theis (2015): <em>Diffusion maps for high-dimensional single-cell analysis of differentiation data</em> Bioinformatics, Vol. 31, Pages 2989–2998, <a href="https://doi.org/10.1093/bioinformatics/btv325">doi:10.1093/bioinformatics/btv325</a></p></li>
<li><p>Haghverdi, Büttner, Wolf, Buettner, Theis (2016): <em>Diffusion pseudotime robustly reconstructs lineage branching</em> Nature Methods, Vol. 13, Pages 845–848</p></li>
<li><p>Angerer, Haghverdi, Büttner, Theis, Marr, Buettner: <em>destiny: diffusion maps for large-scale single-cell data in R</em>, Bioinformatic, Vol. 32, Pages 1241-1243, <a href="https://doi.org/10.1093/bioinformatics/btv715">doi:10.1093/bioinformatics/btv715</a></p></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>