<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Mathematics of single-cell omics - A simple analysis with Seurat</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">A simple analysis with Seurat</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="seurat" class="level3">
<h3 class="anchored" data-anchor-id="seurat">Seurat</h3>
<p>To get a first feel for single-cell RNA-Seq data, we will perform a simple standard analysis using <em>Seurat</em>, an R package offering comprehensive functionality to analyse such data. Seurat has been developed by Rahul Satija and his research group at New York University and is described in multiple publications. The package is named for French pointilist painter Georges Seurat. The web site for Seurat is <a href="https://satijalab.org/seurat/">here</a>.</p>
<p>We load the package in R. We will also need the Matrix package and the magrittr package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( Seurat )</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( Matrix )</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( magrittr ) })</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="first-example-data-pbmcs" class="level3">
<h3 class="anchored" data-anchor-id="first-example-data-pbmcs">First example data: PBMCs</h3>
<p>Our first example data set is a single-cell RNA-Seq data set produced by 10X (the manufacturer of the leading microfluidics platform for single-cell sequencing) to demonstrate their product.</p>
<p>They took a sample of peripheral blood (i.e., blood taken from a vein) of a human donor and removed all red blood cells and platelets (which both have no nucleus) and all white blood cells with multiple nuclei, leaving us only those types of white blood cells that have a single nucleus, i.e., with “peripheral-blood mononuclear cells”: PBMC.</p>
<p>A count matrix has been obtained from the sequencing reads, using 10X’s “Cellranger” software. The matrix is available from teh 10X web site, <a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">here</a>.</p>
<p>The matrix is provided in 10X’s own format. Seurat has a function to load this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>count_matrix <span class="ot">&lt;-</span> <span class="fu">Read10X</span>( <span class="st">"data/pbmc3k/filtered_gene_bc_matrices/hg19/"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="first-look-at-the-count-matrix" class="level4">
<h4 class="anchored" data-anchor-id="first-look-at-the-count-matrix">First look at the count matrix</h4>
<p>The matrix is provided as a sparse matrix (as defined in the Matrix package):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>( count_matrix )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "dgCMatrix"
attr(,"package")
[1] "Matrix"</code></pre>
</div>
</div>
<p>Specifically, it is a column-sparse matrix (one of three standard storage formats for sparse matrices).</p>
<p>Let’s have a look at the top left corner of the matrix</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>count_matrix[ <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span> ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>10 x 10 sparse Matrix of class "dgCMatrix"</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>  [[ suppressing 10 column names 'AAACATACAACCAC-1', 'AAACATTGAGCTAC-1', 'AAACATTGATCAGC-1' ... ]]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                                 
MIR1302-10    . . . . . . . . . .
FAM138A       . . . . . . . . . .
OR4F5         . . . . . . . . . .
RP11-34P13.7  . . . . . . . . . .
RP11-34P13.8  . . . . . . . . . .
AL627309.1    . . . . . . . . . .
RP11-34P13.14 . . . . . . . . . .
RP11-34P13.9  . . . . . . . . . .
AP006222.2    . . . . . . . . . .
RP4-669L17.10 . . . . . . . . . .</code></pre>
</div>
</div>
<p>All entiries are zero (denoted as dot).</p>
<p>We have rows for the genes and columns for the cell barcodes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>( count_matrix )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32738  2700</code></pre>
</div>
</div>
<p>The rows are labelled with gene symbols:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(count_matrix) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "MIR1302-10"    "FAM138A"       "OR4F5"         "RP11-34P13.7" 
 [5] "RP11-34P13.8"  "AL627309.1"    "RP11-34P13.14" "RP11-34P13.9" 
 [9] "AP006222.2"    "RP4-669L17.10" "OR4F29"        "RP4-669L17.2" 
[13] "RP5-857K21.15" "RP5-857K21.1"  "RP5-857K21.2"  "RP5-857K21.3" 
[17] "RP5-857K21.4"  "RP5-857K21.5"  "OR4F16"        "RP11-206L10.3"</code></pre>
</div>
</div>
<p>The columns are labelled with cell barcodes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(count_matrix) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "AAACATACAACCAC-1" "AAACATTGAGCTAC-1" "AAACATTGATCAGC-1" "AAACCGTGCTTCCG-1"
 [5] "AAACCGTGTATGCG-1" "AAACGCACTGGTAC-1" "AAACGCTGACCAGT-1" "AAACGCTGGTTCTT-1"
 [9] "AAACGCTGTAGCCA-1" "AAACGCTGTTTCTG-1" "AAACTTGAAAAACG-1" "AAACTTGATCCAGA-1"
[13] "AAAGAGACGAGATA-1" "AAAGAGACGCGAGA-1" "AAAGAGACGGACTT-1" "AAAGAGACGGCATT-1"
[17] "AAAGATCTGGGCAA-1" "AAAGCAGAAGCCAT-1" "AAAGCAGATATCGG-1" "AAAGCCTGTATGCG-1"</code></pre>
</div>
</div>
<p>The column sums of the matrix tell us how many reads we got per cell:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(count_matrix) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 
            2421             4903             3149             2639 
AAACCGTGTATGCG-1 AAACGCACTGGTAC-1 AAACGCTGACCAGT-1 AAACGCTGGTTCTT-1 
             981             2164             2176             2260 
AAACGCTGTAGCCA-1 AAACGCTGTTTCTG-1 AAACTTGAAAAACG-1 AAACTTGATCCAGA-1 
            1276             1103             3918             2392 
AAAGAGACGAGATA-1 AAAGAGACGCGAGA-1 AAAGAGACGGACTT-1 AAAGAGACGGCATT-1 
            2412             3034             1152              792 
AAAGATCTGGGCAA-1 AAAGCAGAAGCCAT-1 AAAGCAGATATCGG-1 AAAGCCTGTATGCG-1 
            1348             1158             4586             2929 </code></pre>
</div>
</div>
<p>Here is a histogram of the total read counts per cell, on a log10 scale</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(count_matrix) <span class="sc">%&gt;%</span> <span class="fu">log10</span>() <span class="sc">%&gt;%</span> <span class="fu">hist</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="seurat_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s pick an arbitrary cell and ask what genes were expressed in this cell:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>cell <span class="ot">&lt;-</span> <span class="dv">1423</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>count_matrix[ , cell ] <span class="sc">%&gt;%</span> <span class="fu">sort</span>(<span class="at">decreasing=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     B2M   MALAT1   TMSB4X    RPL11    RPS18    RPL13    RPL10   RPL18A 
      72       65       46       25       23       20       19       19 
  RPL13A   MT-CO1    RPS12   RPL23A     RPL3    RPS14     RPS6  PTPRCAP 
      19       19       18       18       18       17       17       17 
    RPS3     RPS2   RPS15A   EEF1A1     RPL7    RPL19    RPL28     RPL9 
      17       17       17       16       16       16       16       15 
   RPS27    RPL32    RPS23    RPL15 HLA-DPB1     RPL6    RPL21     ACTB 
      14       14       14       13       13       13       13       12 
    NKG7     RPS8    HLA-C    RPL12    RPLP1   RPS27A    RPL31    RPL34 
      12       11       11       11       11       10       10       10 
   RPS3A    RPS20    RPLP2     CCL5    RPS15    RPL36   MT-CO2     RPS7 
      10       10       10       10       10        9        9        8 
  RPL35A    RPS4X 
       8        8 </code></pre>
</div>
</div>
<p>We can also ask how often each value appears:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>( count_matrix[ , cell] )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    0     1     2     3     4     5     6     7     8     9    10    11    12 
31919   594    89    30    12    10    15    11    11     2     8     4     2 
   13    14    15    16    17    18    19    20    23    25    46    65    72 
    4     3     1     4     6     3     4     1     1     1     1     1     1 </code></pre>
</div>
</div>
<p>What are the most strongly expressed genes? Let’s average over all cells:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowMeans</span>(count_matrix) <span class="sc">%&gt;%</span> <span class="fu">sort</span>(<span class="at">decreasing=</span><span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  MALAT1   TMSB4X      B2M    RPL10    RPL13   RPL13A      FTL     RPS2 
59.88333 46.00370 44.94926 32.78407 28.55963 28.46037 27.66741 24.13148 
    RPS6     FTH1    RPS18    RPL11    RPL32     RPS3    RPL19    RPS12 
23.16519 21.23741 20.52333 19.12370 19.10815 18.89815 18.15852 17.87259 
   RPLP1     ACTB     RPL3    RPS27    RPS19    RPS14   EEF1A1   RPL18A 
17.57407 17.54185 17.48593 16.70148 16.24370 16.04000 15.52000 15.17741 
  MT-CO1    RPL28   TMSB10    RPS4X    RPLP2    RPL21 
14.51519 14.30741 13.81704 13.76704 13.21630 12.96333 </code></pre>
</div>
</div>
<p>How similar are two arbitrarily picked cells?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>cell1 <span class="ot">&lt;-</span> <span class="dv">1352</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>cell2 <span class="ot">&lt;-</span> <span class="dv">762</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( </span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jitter</span>( count_matrix[,cell1] <span class="sc">+</span> <span class="dv">1</span> ), </span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">jitter</span>( count_matrix[,cell2] <span class="sc">+</span> <span class="dv">1</span>), </span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cex=</span>.<span class="dv">3</span>, <span class="at">log=</span><span class="st">"xy"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="seurat_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="576"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="analysis-with-seurat" class="level3">
<h3 class="anchored" data-anchor-id="analysis-with-seurat">Analysis with Seurat</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>count_matrix <span class="sc">%&gt;%</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">CreateSeuratObject</span>() <span class="sc">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="fu">RunPCA</span>( <span class="at">npcs=</span><span class="dv">20</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="fu">FindNeighbors</span>( <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="fu">FindClusters</span>( <span class="at">resolution=</span><span class="fl">0.5</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="fu">RunUMAP</span>( <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span> ) <span class="ot">-&gt;</span> seu</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  CST3, TYROBP, LST1, AIF1, FTL, FCN1, LYZ, FTH1, S100A9, FCER1G 
       TYMP, CFD, LGALS1, S100A8, CTSS, LGALS2, SERPINA1, SPI1, IFITM3, PSAP 
       CFP, SAT1, IFI30, COTL1, S100A11, NPC2, LGALS3, GSTP1, PYCARD, NCF2 
Negative:  MALAT1, LTB, IL32, CD2, ACAP1, STK17A, CTSW, CD247, CCL5, GIMAP5 
       AQP3, GZMA, CST7, TRAF3IP3, MAL, HOPX, ITM2A, GZMK, MYC, GIMAP7 
       BEX2, ETS1, LDLRAP1, ZAP70, LYAR, RIC3, TNFAIP8, NKG7, KLRG1, SAMD3 
PC_ 2 
Positive:  NKG7, PRF1, CST7, GZMA, GZMB, FGFBP2, CTSW, GNLY, GZMH, SPON2 
       CCL4, FCGR3A, CCL5, CD247, XCL2, CLIC3, AKR1C3, SRGN, HOPX, CTSC 
       TTC38, S100A4, ANXA1, IL32, IGFBP7, ID2, ACTB, XCL1, APOBEC3G, SAMD3 
Negative:  CD79A, MS4A1, TCL1A, HLA-DQA1, HLA-DRA, HLA-DQB1, LINC00926, CD79B, HLA-DRB1, CD74 
       HLA-DPB1, HLA-DMA, HLA-DQA2, HLA-DRB5, HLA-DPA1, HLA-DMB, FCRLA, HVCN1, LTB, BLNK 
       KIAA0125, P2RX5, IRF8, IGLL5, SWAP70, ARHGAP24, SMIM14, PPP1R14A, FCRL2, C16orf74 
PC_ 3 
Positive:  PPBP, PF4, SDPR, SPARC, GNG11, NRGN, GP9, RGS18, TUBB1, CLU 
       HIST1H2AC, AP001189.4, ITGA2B, CD9, TMEM40, CA2, PTCRA, ACRBP, MMD, NGFRAP1 
       TREML1, F13A1, RUFY1, SEPT5, MPP1, TSC22D1, CMTM5, RP11-367G6.3, MYL9, GP1BA 
Negative:  HLA-DQA1, CD79A, CD79B, HLA-DQB1, HLA-DPB1, CD74, HLA-DPA1, MS4A1, HLA-DRB1, HLA-DRB5 
       HLA-DRA, HLA-DQA2, TCL1A, LINC00926, HLA-DMB, HLA-DMA, HVCN1, FCRLA, IRF8, BLNK 
       KIAA0125, SMIM14, PLD4, P2RX5, IGLL5, SWAP70, LAT2, TMSB10, IGJ, MZB1 
PC_ 4 
Positive:  HLA-DQA1, HIST1H2AC, PF4, CD79A, SDPR, CD79B, PPBP, GNG11, HLA-DQB1, SPARC 
       MS4A1, CD74, GP9, HLA-DPB1, RGS18, NRGN, PTCRA, CD9, HLA-DQA2, AP001189.4 
       CLU, TUBB1, CA2, HLA-DRB1, HLA-DPA1, ITGA2B, HLA-DRA, TCL1A, TMEM40, ACRBP 
Negative:  VIM, S100A8, S100A6, S100A4, TMSB10, S100A9, IL32, GIMAP7, S100A10, LGALS2 
       RBP7, MAL, FCN1, LYZ, CD2, S100A12, MS4A6A, FYB, S100A11, AQP3 
       GIMAP4, FOLR3, ANXA1, MALAT1, AIF1, GIMAP5, IL8, IFI6, TRABD2A, ASGR1 
PC_ 5 
Positive:  GZMB, FGFBP2, NKG7, GNLY, PRF1, CCL4, CST7, SPON2, GZMA, GZMH 
       CLIC3, XCL2, CTSW, TTC38, AKR1C3, CCL5, IGFBP7, XCL1, S100A8, CCL3 
       TYROBP, HOPX, CD160, HAVCR2, S100A9, FCER1G, PTGDR, LGALS2, RBP7, S100A12 
Negative:  LTB, VIM, AQP3, PPA1, MAL, KIAA0101, CD2, CORO1B, CYTIP, FYB 
       IL32, TRADD, ANXA5, TUBA1B, HN1, PTGES3, TYMS, ITM2A, COTL1, GPR183 
       ACTG1, TNFAIP8, ATP5C1, TRAF3IP3, GIMAP4, PRDX1, ZWINT, ABRACL, NGFRAP1, LDLRAP1 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 2700
Number of edges: 120276

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8640
Number of communities: 8
Elapsed time: 0 seconds</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:24:27 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:24:27 Read 2700 rows and found 20 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:24:27 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:24:27 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
18:24:27 Writing NN index file to temp file /tmp/Rtmpmb1iyX/file1446437f4addf
18:24:27 Searching Annoy index using 1 thread, search_k = 3000
18:24:28 Annoy recall = 100%
18:24:29 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
18:24:29 Initializing from normalized Laplacian + noise (using RSpectra)
18:24:30 Commencing optimization for 500 epochs, with 111204 positive edges
18:24:34 Optimization finished</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>( seu ) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="seurat_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In this plot, each cell is represented by a point in a so-called dimension-reduced embedding. This means that the points have been arranged such that cells that are similar appear close to each other while cells that are very different are put far apart from each other.</p>
<p>We will discuss soon in detail how such dimension reduction works and, importantly, what is meant by “similar”.</p>
<p>Seurat has also run a “clustering” algorithm that assigns the cells to different “clusters” of cells that are similar to each other. Colours are used here to indicate cluster membership.</p>
<p>We observe that the cells fall into three big groups that each split into a number of smaller clusters.</p>
<section id="identifying-cell-types" class="level4">
<h4 class="anchored" data-anchor-id="identifying-cell-types">Identifying cell types</h4>
<p>We suspect that these represent different types of white blood cells. There should be B-cells, T-cells and perhaps monocytes.</p>
<p>T cells are defined as white blood cells that show a the T-cell receptor as their surface, a protein complex made up of various proteins, one of which is called the T-cell receptor epsilon chain. This protein is described for the gene CD3E. We can ask Seurat to indicate the read counts for this gene in the UMAP plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>( seu, <span class="st">"CD3E"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="seurat_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We conclede that clusters 0 and 3 are probably the T cells.</p>
<p>We can also colour for CD79A, a marker for B cells:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>( seu, <span class="st">"CD79A"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="seurat_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To find out what the group to the left is, we can go another way. Let’s ask Seurat to find genes that are expressed much mroe strongly in clusters 1 and 4 than in clusters 0, 2, 3, 5:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FindMarkers</span>( seu, <span class="at">ident.1=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">4</span>), <span class="at">ident.2=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">5</span>), <span class="at">test.use=</span><span class="st">"t"</span> ) <span class="ot">-&gt;</span> m</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>( m, <span class="dv">50</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 p_val avg_log2FC pct.1 pct.2     p_val_adj
LST1      0.000000e+00  5.2602174 0.969 0.147  0.000000e+00
CST3      0.000000e+00  5.6816322 0.994 0.193  0.000000e+00
TYROBP    0.000000e+00  4.5599105 0.995 0.194  0.000000e+00
FCER1G    0.000000e+00  3.9303734 0.953 0.153  0.000000e+00
AIF1      0.000000e+00  4.8394973 0.968 0.171  0.000000e+00
PTPRCAP   0.000000e+00 -4.3390790 0.123 0.820  0.000000e+00
LGALS1    0.000000e+00  3.5888431 0.982 0.303  0.000000e+00
CTSS      0.000000e+00  3.3704116 0.956 0.370  0.000000e+00
IL32      0.000000e+00 -4.5060681 0.123 0.703  0.000000e+00
S100A11   0.000000e+00  2.9021375 0.959 0.381  0.000000e+00
LYZ       0.000000e+00  5.5523830 0.995 0.471  0.000000e+00
SAT1      0.000000e+00  3.1880331 0.972 0.454  0.000000e+00
COTL1     0.000000e+00  2.9140976 0.974 0.505  0.000000e+00
S100A6    0.000000e+00  2.4265040 0.992 0.712  0.000000e+00
S100A4    0.000000e+00  2.1051825 1.000 0.752  0.000000e+00
OAZ1      0.000000e+00  1.9165328 0.995 0.877  0.000000e+00
FTL       0.000000e+00  3.6241681 1.000 0.983  0.000000e+00
FTH1      0.000000e+00  3.1675076 1.000 0.984  0.000000e+00
CD3D     4.886112e-319 -4.3674643 0.092 0.670 1.599615e-314
TYMP     3.882155e-306  4.2566370 0.926 0.151 1.270940e-301
S100A9   1.783818e-297  7.0670428 0.962 0.154 5.839865e-293
PSAP     1.902707e-294  2.9956639 0.934 0.359 6.229083e-290
FCN1     9.535250e-274  5.5804410 0.896 0.092 3.121650e-269
CYBA     8.453508e-272  1.5333488 0.991 0.797 2.767510e-267
MALAT1   1.181579e-262 -1.5311747 1.000 1.000 3.868252e-258
LTB      4.335583e-260 -3.5012014 0.348 0.824 1.419383e-255
NPC2     1.360721e-240  3.2405179 0.896 0.252 4.454729e-236
CD3E     8.066458e-236 -3.7395374 0.104 0.612 2.640797e-231
CFD      7.012340e-210  5.9290434 0.822 0.036 2.295700e-205
S100A8   1.234624e-205  7.2702939 0.855 0.084 4.041912e-201
HLA-DRA  9.958375e-200  0.9548420 0.945 0.423 3.260173e-195
GABARAP  3.036143e-190  2.4215683 0.881 0.351 9.939724e-186
CD7      2.761808e-188 -4.2406042 0.058 0.475 9.041608e-184
GSTP1    8.906341e-186  2.9200924 0.877 0.315 2.915758e-181
RPS27    6.463844e-178 -1.2827741 0.988 0.998 2.116133e-173
CD68     5.216417e-175  5.1623519 0.757 0.039 1.707751e-170
GPX1     2.638599e-174  3.1465728 0.853 0.267 8.638245e-170
SERPINA1 5.945946e-174  4.6390651 0.770 0.069 1.946584e-169
LCK      2.819845e-170 -3.6083579 0.058 0.464 9.231609e-166
NEAT1    1.968831e-169  1.8285530 0.928 0.540 6.445560e-165
RPL3     5.794248e-169 -1.1390928 0.994 0.999 1.896921e-164
PYCARD   8.652551e-167  2.7349426 0.824 0.215 2.832672e-162
LGALS2   4.290216e-165  6.6287302 0.743 0.029 1.404531e-160
RPL13A   9.270560e-165 -0.8866787 1.000 1.000 3.034996e-160
RPS27A   2.750106e-158 -1.7498987 0.911 0.996 9.003298e-154
CD2      8.578003e-155 -4.2838412 0.052 0.407 2.808267e-150
RAC1     1.921305e-154  2.2617174 0.845 0.294 6.289968e-150
CXCR4    2.275604e-153 -2.6822981 0.285 0.701 7.449873e-149
RPL23A   2.157301e-152 -1.2329204 0.980 0.997 7.062571e-148
TKT      2.978425e-152  2.8123604 0.793 0.200 9.750769e-148</code></pre>
</div>
</div>
<p>From this list, an immunologist would quickly see that these cells are monocytes.</p>
</section>
<section id="back-to-math" class="level4">
<h4 class="anchored" data-anchor-id="back-to-math">Back to math</h4>
<p>What is happening inside Seurat.</p>
<p>A few questions we might have:</p>
<ul>
<li>How do we define “similarity” of cells?</li>
<li>Each cell is represented by a vector of count values. Can we simply define a metric or distance measure in that space to quantify similarity?</li>
<li>It will turn out that this does not work, due to the counting noise. Why?</li>
<li>How can we make cells with differing total read count comparable?</li>
<li>We will overcome the count-noise issue by performing a principal component analysis (a matrix decomposition based on the eigendecomposition). Why does that work?</li>
<li>The cells seem to be living in a high-dimensional space. Can this really be faithfully shown by a two-dimensional plot? How do we check the faithfulness? Here, we will get into a rabbit hole involving multi-variate optimization, metrics in distribution spaces, interactive visualizations and maybe GPU programming.</li>
<li>How to suitable define “clusters” of similar cells? This will lead us to applications of graph theory.</li>
</ul>
</section>
<section id="the-manifold-hypothesis" class="level4">
<h4 class="anchored" data-anchor-id="the-manifold-hypothesis">The manifold hypothesis</h4>
<p>A teaser to the core idea in single-cell analysis:</p>
<p>Cells “live” in a “state space” that is spanned by all the possible configurations of the cells in the organism. When a cell reacts to a stimulus or changes its state for another reason, this will be a continuous (maybe even smooth?) move in this statespace.</p>
<p>However, we cannot observe the state space directly – but we assume that each state is associated with a transcriptome, i.e., a measure or distribution in the space of read count vectors spanned by the genes (i.e., of possible columns in the count matrix).</p>
<p>We assume that the distribution of the cells in our sample lives on a <em>manifold</em> in state space, and we wish to learn first the topology of this manifold, then gain a metric and an atlas for it, then infer the distribution/measure on this manifold, from which our cells have been sampled. And finally, we have to translate this mathematical description into a biological insights.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>