<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Mathematics of single-cell omics - Trajectories</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="fullcontent">

<div id="quarto-search-results"></div>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Trajectories</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="standard-analysis" class="level3">
<h3 class="anchored" data-anchor-id="standard-analysis">Standard analysis</h3>
<p>We load our usual example data</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-1_0ed53809a9007234ab1a2848db7f73d0">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( tidyverse )</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( Matrix )</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( sparseMatrixStats )</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>( Seurat ) })</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ReadMtx</span>( <span class="st">"~/Downloads/ifnagrko/ifnagrko_raw_counts.mtx.gz"</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/ifnagrko/ifnagrko_obs.csv"</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"~/Downloads/ifnagrko/ifnagrko_var.csv"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">cell.sep=</span><span class="st">","</span>, <span class="at">feature.sep=</span><span class="st">","</span>, <span class="at">skip.cell=</span><span class="dv">1</span>, <span class="at">skip.feature=</span><span class="dv">1</span>, </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">mtx.transpose=</span><span class="cn">TRUE</span>) <span class="ot">-&gt;</span> count_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-2_46812a47bdd1a2dbbe34c3cb17cf33ac">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>count_matrix <span class="sc">%&gt;%</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">CreateSeuratObject</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FindVariableFeatures</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">RunPCA</span>( <span class="at">npcs=</span><span class="dv">20</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">FindNeighbors</span>( <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">FindClusters</span>( <span class="at">resolution=</span><span class="fl">0.5</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">RunUMAP</span>( <span class="at">dims=</span><span class="dv">1</span><span class="sc">:</span><span class="dv">20</span> ) <span class="ot">-&gt;</span> seu</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  Apoe, Aldoc, Sparcl1, Sdc4, Ptn, Cmtm5, Glul, Gpr37l1, Fxyd1, Atp1b2 
       S100a1, Slc4a4, Slc1a3, Prxl2a, F3, Itm2b, Mt1, Rgcc, Prdx6, Sfxn5 
       Sat1, Scrg1, Dbi, Hes5, Luzp2, Plaat3, Pla2g7, Sash1, Plpp3, Sparc 
Negative:  Tubb5, Sox11, Tubb3, Stmn1, Jpt1, Hmgb3, Ptma, Sox4, Dlx2, Cd24a 
       Igfbpl1, Dlx6os1, Map1b, Stmn2, Abracl, Tmsb4x, Lmnb1, Cdca7, Ccnd2, Elavl4 
       Cdk4, Dcx, Arx, Uchl1, EYFP, Celf4, Dlx5, Nrxn3, H1fx, Hmgn2 
PC_ 2 
Positive:  Ctss, C1qc, Laptm5, Csf1r, Trem2, C1qa, Cx3cr1, C1qb, Tyrobp, Ly86 
       Fcer1g, Siglech, Selplg, Fcrls, Tmem119, Fcgr3, Apbb1ip, Unc93b1, Cd53, Lpcat2 
       Spi1, Pld4, Olfml3, Irf8, Ctsh, Aif1, Cd300c2, Fyb, Otulinl, Mylip 
Negative:  Rorb, Cldn10, Clu, Mt3, Ntsr2, Mfge8, S1pr1, Id4, Slc1a2, Acsl6 
       Plpp3, Sox9, Ddah1, Bcan, Cxcl14, Btbd17, Mlc1, Cspg5, Fjx1, Aqp4 
       Ntm, Acsl3, Gabrb1, Tspan7, Lsamp, Chst2, Mt2, Lhx2, Slc39a12, Glud1 
PC_ 3 
Positive:  Atp1a3, Camk2b, Snhg11, Syt1, Nrip3, Kcnj4, Scg2, Snap25, Dnm1, Pcp4 
       Icam5, Ndrg4, Eef1a2, Eno2, Ano3, Ryr2, Arpp21, Ptk2b, Gng4, Kcna4 
       Penk, Slc4a10, Snca, Gad1, Rprml, Grin2a, C1qtnf4, Shisa8, Camk2a, Kcnb2 
Negative:  Hmgb2, Top2a, Pbk, Birc5, Mki67, Cdk1, Cdca8, Spc24, Cenpf, Spc25 
       Prc1, Rrm2, Mdk, Nusap1, Tpx2, Cdca3, Knl1, Ckap2l, Esco2, Aurkb 
       Cenpm, Ccna2, Bub1, Cks2, Kif11, Hist1h3c, Hist1h1b, Hmmr, Pclaf, Fbxo5 
PC_ 4 
Positive:  C1qc, C1qa, Ctss, Trem2, Csf1r, C1qb, Cx3cr1, Laptm5, Fcer1g, Tyrobp 
       Ly86, Siglech, Selplg, Fcrls, Fcgr3, Hexb, Spi1, Cd53, Itgb5, Pld4 
       Ptgs1, Cd300c2, Aif1, Irf8, Fyb, Itgam, Cyth4, Ltc4s, Otulinl, Cd37 
Negative:  Frzb, Apod, Npy, Plp1, Vtn, Foxd3, Wnt6, Nr2f2, Edil3, Sox10 
       Gsn, Matn4, Fbln2, Aspa, Aqp1, Igf1, Plat, Lpar1, Igfbp4, Erbb3 
       Fabp7, Plppr4, Ptgds, Col23a1, Alx3, Hey2, Cd59a, Fam3c, Scd1, Mybpc1 
PC_ 5 
Positive:  Stmn2, Igfbpl1, Cd24a, Nrep, Sox4, Map1b, Stmn4, Tubb3, Shtn1, Dlx6os1 
       Dcx, Ly6h, Sox11, Jpt1, Mpped2, Stmn1, Plxna4, Pbx3, Elavl4, Uchl1 
       Runx1t1, Cald1, Foxp2, Dlx2, Gad2, Celf4, Pfn2, Dlx5, Sp8, Tubb5 
Negative:  Top2a, Pbk, Birc5, Mki67, Spc25, Cdk1, Prc1, Nusap1, Spc24, Esco2 
       Tpx2, Knl1, Aurkb, Cenpf, Cdca8, Ckap2l, Kif11, Cdca3, Hist1h3c, Hmmr 
       Ccna2, Bub1, Incenp, Hist1h2af, Ndc80, Cit, Fbxo5, Kif4, Sgo1, Kif22 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 18302
Number of edges: 616069

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9191
Number of communities: 19
Elapsed time: 5 seconds</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:04:48 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:04:48 Read 18302 rows and found 20 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:04:48 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>12:04:48 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
12:04:51 Writing NN index file to temp file /tmp/Rtmp4gTY5z/file33b7e39258597
12:04:51 Searching Annoy index using 1 thread, search_k = 3000
12:05:01 Annoy recall = 100%
12:05:01 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
12:05:02 Initializing from normalized Laplacian + noise (using RSpectra)
12:05:07 Commencing optimization for 200 epochs, with 776086 positive edges
12:05:18 Optimization finished</code></pre>
</div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-3_56e8ac8b5fbd8f311cb476bf68f80687">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">UMAPPlot</span>( seu, <span class="at">label=</span><span class="cn">TRUE</span> ) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This time, we will concetrate on the long elongated main structure, which we will call the “lineage” in the following. It is a snapshot of the development of astrocytes that act as neural stem cells and become transient amplifying progenitors (TAPs) which undergo cell cycle, i.e., divide and multiply, and the turn into neuroblasts and finally neurons.</p>
<p>To orient us in the plot, we highlight the expression of Aqp4 (aquaporin-4, a marker for astrocytes), Mki67 (a marker for prliferating, i.e., dividing cells), Dcx (doublecortin, a marker for neuroblasts) and Gria1 (Glutamate ionotropic receptor, AMPA type, subunit 1; a marker for mature neurons)</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-4_659b756cad37dfd481529e314f14df12">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>( seu, <span class="fu">c</span>( <span class="st">"Aqp4"</span>, <span class="st">"Mki67"</span>, <span class="st">"Dcx"</span>, <span class="st">"Gria1"</span> ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We conclude that the lineage is well covered by the following clusters</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-5_1d61e6943730ea340037e9a17055aded">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>lineage_clusters <span class="ot">&lt;-</span>  <span class="fu">c</span>( <span class="dv">10</span>, <span class="dv">9</span>, <span class="dv">0</span>, <span class="dv">13</span>, <span class="dv">14</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">11</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">7</span> ) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Just out of curiosity, we also try to identify clusters 4 and 8:</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-6_0d246930e9692030c10ad8b237cea633">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>presto<span class="sc">::</span><span class="fu">wilcoxauc</span>(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">LayerData</span>(seu),</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">factor</span>(<span class="fu">case_when</span>(</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> lineage_clusters <span class="sc">~</span> <span class="st">"lineage"</span>,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> <span class="fu">c</span>( <span class="dv">4</span>, <span class="dv">8</span> ) <span class="sc">~</span> <span class="st">"4_or_8"</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"other"</span> )) ) <span class="ot">-&gt;</span> wa</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(wa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  feature  group     avgExpr         logFC statistic       auc         pval
1    Xkr4 4_or_8 0.003235569 -0.0303111155  16421626 0.4862780 1.501567e-12
2     Rp1 4_or_8 0.000000000 -0.0004121176  16877733 0.4997842 3.431328e-01
3   Sox17 4_or_8 0.000000000 -0.0098443771  16794453 0.4973181 8.092744e-04
4 Gm37323 4_or_8 0.000000000 -0.0006945437  16872528 0.4996301 2.144507e-01
5  Mrpl15 4_or_8 0.080757542 -0.0771071674  15531646 0.4599238 2.285892e-20
6  Lypla1 4_or_8 0.023977097  0.0007956678  16912116 0.5008024 6.802492e-01
          padj    pct_in     pct_out
1 5.070133e-12 0.4322767  3.16892725
2 4.356080e-01 0.0000000  0.04315660
3 1.635191e-03 0.0000000  0.53637485
4 2.968308e-01 0.0000000  0.07398274
5 9.783259e-20 9.4140250 17.29963009
6 7.320138e-01 3.0259366  2.86066584</code></pre>
</div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-7_1109ee201e12a4ced644ea778253096d">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>presto<span class="sc">::</span><span class="fu">top_markers</span>( wa )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
    rank `4_or_8` lineage other  
   &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;  
 1     1 Fabp7    Tubb2b  Rhob   
 2     2 Apod     Meis2   Bin1   
 3     3 Npy      Fabp5   Trf    
 4     4 Ptn      Pfn2    Sgk1   
 5     5 Frzb     Gpm6a   Serinc3
 6     6 Plp1     Ccdc88a Abcg1  
 7     7 Gsn      Pantr1  Jund   
 8     8 Nr2f2    Pou3f2  Jun    
 9     9 Matn4    Foxg1   Smad7  
10    10 Itm2b    Gnao1   H2-D1  </code></pre>
</div>
</div>
<p>As a try, I’ve asked ChatGPT what these genes point to and it <a href="https://chatgpt.com/share/f2a5e86b-9909-4c50-8b6b-798f9596df71">replied</a> that the cells in our clusters 4 and 8 are oligondendrocyte precursor cells (OPCs). This matches my expectation.</p>
</section>
<section id="aim-trajectory" class="level3">
<h3 class="anchored" data-anchor-id="aim-trajectory">Aim: Trajectory</h3>
<p>Our first aim for today is to fit a “pseudotime trejectory” to the lineage. This means thatw e want to assign to each cell in the lineage a real number, which we call it “pseudotime” that monotonally increases along the putative developmental trajectory from astrocytic neural stem cells via TAPs and neuroblasts to neurons.</p>
<p>We will do this by fitting a “principal curve”, i.e.&nbsp;a curve in PCA space that tracks along the lineage and is fitted such that the squared sum of the cells’ distance to their respectively closest point on the curve (their projection image) is minimal. The distance of this projection image to the curve start (measured along the curve) will be used as pseudotime.</p>
<p>The “principal curve” method is described in detail in <a href="principal_curves.html">this section of the lecture notes</a>.</p>
<p>Here, we will use the function from the <code>princurve</code> package.</p>
<p>As preparation, we first explore distances with sleepwalk:</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-8_76dfba74dabb711b9706bbfe7f69ef22">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sleepwalk<span class="sc">::</span><span class="fu">sleepwalk</span>( <span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>), <span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We notice that the cycling cells have a lot of distance to the non-cycling lineage cells. This will cause problems because the principal curve alorithm cannot deal with loops, i.e., the curve should pass besides the cell-cycle loop. However, the distances to the curve will then become large right in the middle, deflecting the curve.</p>
<p>Therefore, let’s exclude clusters 6 and 11:</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-9_1c4cd0fc533607217de4ebc091844355">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>lineage_clusters_2 <span class="ot">&lt;-</span> <span class="fu">setdiff</span>( lineage_clusters, <span class="fu">c</span>( <span class="dv">6</span>, <span class="dv">11</span> ) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="fitting-the-principal-curve" class="level3">
<h3 class="anchored" data-anchor-id="fitting-the-principal-curve">Fitting the principal curve</h3>
<p>Now, we could the <code>princomp</code> package to fit the principal curve</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-10_ddeaecb2571f48a4ea0a63c5fc9391c6">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>princurve<span class="sc">::</span><span class="fu">principal_curve</span>( </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>)[ seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> lineage_clusters_2, ], </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">df=</span><span class="dv">10</span>, <span class="at">trace=</span><span class="cn">TRUE</span>, <span class="at">approx_points=</span><span class="dv">1000</span> ) <span class="ot">-&gt;</span> prc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting curve---distance^2: 38216203191
Iteration 1---distance^2: 777432.5
Iteration 2---distance^2: 595846
Iteration 3---distance^2: 569788.5
Iteration 4---distance^2: 558439.6
Iteration 5---distance^2: 549200.9
Iteration 6---distance^2: 544509.5
Iteration 7---distance^2: 542890.9
Iteration 8---distance^2: 541627.2
Iteration 9---distance^2: 538015.4
Iteration 10---distance^2: 524569.8</code></pre>
</div>
</div>
<p>This function returns for each cell a pseudotime value <code>lambda</code> and a projected position on the curve in PCA space, <code>s</code>.</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-11_57e002fc9a099cf800c1463dbc04c6e9">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>( <span class="at">rownames=</span><span class="st">"cell"</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>( <span class="fu">enframe</span>( prc<span class="sc">$</span>lambda, <span class="st">"cell"</span>, <span class="st">"lambda"</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>lambda ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>() <span class="sc">+</span> <span class="fu">scale_color_viridis_c</span>(<span class="at">option=</span><span class="st">"D"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(cell)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can assign a pseudotime to the remaining cells by finding the closest curve point:</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-12_6eeb9a7e37fa2eeb77712adf3fb34336">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>FNN<span class="sc">::</span><span class="fu">get.knnx</span>( prc<span class="sc">$</span>s, <span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>), <span class="dv">1</span> ) <span class="ot">-&gt;</span> nnres</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>prc<span class="sc">$</span>lambda[ nnres<span class="sc">$</span>nn.index[,<span class="dv">1</span>] ] <span class="sc">%&gt;%</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  { ( <span class="fu">max</span>(.) <span class="sc">-</span> . ) <span class="sc">/</span> <span class="fu">max</span>(.) } <span class="sc">%&gt;%</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_names</span>( <span class="fu">rownames</span>(<span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>)) ) <span class="ot">-&gt;</span> seu<span class="sc">$</span>pt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Her we have rescaled the pseudotime to [0;1] and also reversed the dirction, so thjat it now increases from stem cells towards neuroblasts.</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-13_062ed1d6ac9badc62955f58d64f0d288">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>( <span class="at">rownames=</span><span class="st">"cell"</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>( <span class="fu">enframe</span>( seu<span class="sc">$</span>pt, <span class="st">"cell"</span>, <span class="st">"pt"</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>pt ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>() <span class="sc">+</span> <span class="fu">scale_color_viridis_c</span>(<span class="at">option=</span><span class="st">"D"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(cell)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We should also check how for each cell is from the curve</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-14_d093140f8bf65bf89cba9f9eeaf5a342">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>nnres<span class="sc">$</span>nn.dist[,<span class="dv">1</span>] <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set_names</span>( <span class="fu">rownames</span>(<span class="fu">Embeddings</span>(seu,<span class="st">"pca"</span>)) ) <span class="ot">-&gt;</span> seu<span class="sc">$</span>dist_to_curve</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="fu">Embeddings</span>(seu,<span class="st">"umap"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">as_tibble</span>( <span class="at">rownames=</span><span class="st">"cell"</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>( <span class="fu">enframe</span>( seu<span class="sc">$</span>dist_to_curve, <span class="st">"cell"</span>, <span class="st">"dist"</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2, <span class="at">col=</span>dist ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_equal</span>() <span class="sc">+</span> <span class="fu">scale_color_viridis_c</span>( <span class="at">option=</span><span class="st">"D"</span>, <span class="at">trans=</span><span class="st">"log10"</span>, <span class="at">direction=</span><span class="sc">-</span><span class="dv">1</span> )  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(cell)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="expression-dynamics" class="level3">
<h3 class="anchored" data-anchor-id="expression-dynamics">Expression dynamics</h3>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-15_2e6f035c654de005fd907f50a25d3bb7">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt =</span> seu<span class="sc">$</span>pt, </span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">dist =</span> seu<span class="sc">$</span>dist_to_curve,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">in_lineage =</span> seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> lineage_clusters,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">expr =</span> <span class="fu">LayerData</span>(seu)[<span class="st">"Slc1a3"</span>,] ) <span class="sc">%&gt;%</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">expr =</span> <span class="fu">ifelse</span>( expr<span class="sc">&gt;</span><span class="dv">0</span>, expr, <span class="fu">runif</span>( <span class="fu">n</span>(), <span class="sc">-</span>.<span class="dv">2</span>, <span class="dv">0</span> ) ) ) <span class="ot">-&gt;</span> tbl</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>tbl <span class="sc">%&gt;%</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>( in_lineage ) <span class="sc">%&gt;%</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>ggplot <span class="sc">+</span></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">x=</span>pt, <span class="at">y=</span>expr, <span class="at">col=</span>dist ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We now fit a smooth curve through this scatte plot.</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-16_3f9d5972443e8d3bdc2de92ceedf68fa">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>( locfit )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>locfit 1.5-9.9   2024-03-01</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'locfit'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    none</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">pt =</span> seu<span class="sc">$</span>pt, </span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">dist =</span> seu<span class="sc">$</span>dist_to_curve,</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">in_lineage =</span> seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> lineage_clusters,</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">count =</span> <span class="fu">LayerData</span>(seu,<span class="st">"count"</span>)[<span class="st">"Slc1a3"</span>,],</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">total =</span> <span class="fu">colSums</span>( <span class="fu">LayerData</span>(seu,<span class="st">"count"</span>) ),</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">expr =</span> <span class="fu">LayerData</span>(seu)[<span class="st">"Slc1a3"</span>,] ) <span class="sc">%&gt;%</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">expr =</span> <span class="fu">ifelse</span>( expr<span class="sc">&gt;</span><span class="dv">0</span>, expr, <span class="fu">runif</span>( <span class="fu">n</span>(), <span class="sc">-</span>.<span class="dv">2</span>, <span class="dv">0</span> ) ) ) <span class="ot">-&gt;</span> tbl</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">locfit</span>( count <span class="sc">~</span> pt, tbl, <span class="at">weight=</span>total, <span class="at">family=</span><span class="st">"poisson"</span> )</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>( <span class="at">pt =</span> <span class="fu">seq</span>( <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out=</span><span class="dv">1000</span> ) ) <span class="sc">%&gt;%</span></span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a><span class="fu">mutate</span>( <span class="at">y =</span> <span class="fu">predict</span>( fit, pt ) ) <span class="ot">-&gt;</span> tbl_fit</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>tbl <span class="sc">%&gt;%</span></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a><span class="fu">filter</span>( in_lineage ) <span class="sc">%&gt;%</span></span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>( <span class="fu">aes</span>( <span class="at">x=</span>pt ) ) <span class="sc">+</span></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>( <span class="fu">aes</span>( <span class="at">y =</span> count<span class="sc">/</span>total <span class="sc">+</span> <span class="fl">1e-4</span>, <span class="at">col=</span>dist ), <span class="at">size=</span>.<span class="dv">3</span> ) <span class="sc">+</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>( <span class="fu">aes</span>( <span class="at">y =</span> y ), <span class="at">data=</span>tbl_fit, <span class="at">col=</span><span class="st">"magenta"</span> ) <span class="sc">+</span></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span> <span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-17_6fad0e2e3de10898a2f15ef3c206f372">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LayerData</span>(seu)[ , seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> lineage_clusters ] <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowVars</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>( <span class="at">decreasing=</span><span class="cn">TRUE</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="dv">20</span>) <span class="sc">%&gt;%</span> <span class="fu">names</span>() <span class="ot">-&gt;</span> genes</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>tg2 <span class="ot">&lt;-</span> <span class="fu">seq</span>( <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out=</span><span class="dv">100</span> )</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>( genes, <span class="cf">function</span>(gene) {</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">&lt;-</span> <span class="fu">locfit</span>( <span class="fu">LayerData</span>(seu,<span class="st">"count"</span>)[gene,] <span class="sc">~</span> seu<span class="sc">$</span>pt, </span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">weight=</span>seu<span class="sc">$</span>nCount_RNA, <span class="at">family=</span><span class="st">"poisson"</span> )</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"."</span>) </span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>( fit, tg2 )</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>} ) <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="ot">-&gt;</span> fits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>....................</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>fits[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              [,1]        [,2]        [,3]        [,4]        [,5]
Cst3   0.035855217 0.037067666 0.037784438 0.037992608 0.037700510
Slc1a2 0.012262572 0.011624325 0.010907536 0.010137774 0.009338986
Mt3    0.008338254 0.008435790 0.008362052 0.008128155 0.007753874
Clu    0.004685951 0.005638626 0.006458559 0.007050345 0.007343839
Mt1    0.009045342 0.008738399 0.008366646 0.007942673 0.007479294</code></pre>
</div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-18_56c35867aed3c33c636ccaae9eb3066f">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>( <span class="fu">t</span>(fits) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-19_1dbeb38fcb948cf56bee9cf97fff4dac">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>fitsz <span class="ot">&lt;-</span> fits <span class="sc">/</span> <span class="fu">rowMaxs</span>(fits)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>( <span class="fu">t</span>(fitsz), <span class="at">yaxt=</span><span class="st">"n"</span> )</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>( <span class="dv">2</span>, <span class="fu">seq</span>( <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out=</span><span class="fu">nrow</span>(fitsz) ), <span class="fu">rownames</span>(fitsz), <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.axis=</span>.<span class="dv">5</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-20_fda830e72696e7120a12f10bf1f76d42">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>fitszs <span class="ot">&lt;-</span> fitsz[ <span class="fu">order</span>( <span class="sc">-</span><span class="fu">apply</span>( fitsz, <span class="dv">1</span>, which.max ) ), ]</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>( <span class="fu">t</span>(fitszs), <span class="at">yaxt=</span><span class="st">"n"</span> )</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">axis</span>( <span class="dv">2</span>, <span class="fu">seq</span>( <span class="dv">0</span>, <span class="dv">1</span>, <span class="at">length.out=</span><span class="fu">nrow</span>(fitszs) ), <span class="fu">rownames</span>(fitszs), <span class="at">las=</span><span class="dv">2</span>, <span class="at">cex.axis=</span>.<span class="dv">5</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="trajectory_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The same now with the top hundred genes:</p>
<div class="cell" data-hash="trajectory_cache/html/unnamed-chunk-21_e602c892cff28ed3fcea6088446df20d">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LayerData</span>(seu)[ , seu<span class="sc">$</span>seurat_clusters <span class="sc">%in%</span> lineage_clusters ] <span class="sc">%&gt;%</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowVars</span>() <span class="sc">%&gt;%</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>( <span class="at">decreasing=</span><span class="cn">TRUE</span> ) <span class="sc">%&gt;%</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="dv">100</span>) <span class="sc">%&gt;%</span> <span class="fu">names</span>() <span class="ot">-&gt;</span> genes</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>( genes, <span class="cf">function</span>(gene) {</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">&lt;-</span> <span class="fu">locfit</span>( <span class="fu">LayerData</span>(seu,<span class="st">"count"</span>)[gene,] <span class="sc">~</span> seu<span class="sc">$</span>pt, </span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">weight=</span>seu<span class="sc">$</span>nCount_RNA, <span class="at">family=</span><span class="st">"poisson"</span> )</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"."</span>) </span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">predict</span>( fit, tg2 )</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>} ) <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="ot">-&gt;</span> fits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>....................................................................................................</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>